File: tests/pages/HomePage.js
/** HomePage class extends BasePage 
 * Represents the RENFE homepage and search functionality
*/
import { BasePage } from "./BasePage.js";
import { getFutureDate } from "../utils/testData.js";

export class HomePage extends BasePage {
    // Locators
    
    get dateOfDepartureInput() {
        return this.page.locator('//input[@id="first-input"]');
    }
    get oneWayRadio() {
        return this.page.locator('//label[contains(text(),"Viaje solo ida")]').first();// span[@class="lightpick__border"]  div[@class="lightpick__trip-header"]
    }
    get oneWayTab() {
        return this.page.locator('//label[@class="lightpick__label"]').first();
    }
    get originStation() {
        return this.page.locator('//input[@id="origin"]');
    }
    get destinationStation() {
        return this.page.locator('//input[@id="destination"]');
    }
    get departureDate() {//departure date input need ot resolve
        return this.page.locator('//input[@class="rf-daterange-picker-alternative__input"]');//input[@class='rf-daterange-picker-alternative__ipt']
    }
    get searchButton() {   
        return this.page.locator('//button[@type="submit"]');
    }
    get focusOnAlertDialog(){
        return this.page.locator('//div[@class="ot-sdk-container"]');
    }
    get acceptCookiesButton() {
        return this.page.locator('//button[@id="onetrust-accept-btn-handler"]').first();// button:has-text("Accept"), button:has-text("Aceptar"), button:has-text("accept all"), [data-testid="accept-cookies"]
    }
    
    /**
     * Accept all cookies
     */
    async acceptCookies() {
        try {
            const cookieButton = this.acceptCookiesButton;
            if (await cookieButton.isVisible({ timeout: 10000 })) {
                // Focus the alert dialog first
                try {
                    const alertDialog = this.focusOnAlertDialog;
                    if (await alertDialog.isVisible({ timeout: 2000 })) {
                        await alertDialog.evaluate((el) => {
                            if (!el.hasAttribute('tabindex')) el.setAttribute('tabindex', '-1');
                            el.focus();
                        });
                        await this.page.waitForTimeout(100);
                    }
                } catch (e) {
                    // ignore if dialog not found
                }

                // Click cookie button with Enter fallback
                try {
                    await cookieButton.click();
                } catch (clickErr) {
                    try {
                        await cookieButton.press('Enter');
                    } catch (pressErr) {
                        await this.page.keyboard.press('Enter');
                    }
                }
                await this.page.waitForLoadState('networkidle');
            }
        } catch (error) {
            // Cookie banner might not be present, continue with test
        }
    }

    /**
     * Wait for the entire page to be fully loaded
     */
    async waitForPageToLoad() {
        // Wait for network to be idle
        await this.page.waitForLoadState('networkidle');
        
        // Wait for key page elements to be visible
        await this.dateOfDepartureInput.waitFor({ state: 'visible', timeout: 10000 });
        await this.originStation.waitFor({ state: 'visible', timeout: 10000 });
        await this.destinationStation.waitFor({ state: 'visible', timeout: 10000 });
        await this.searchButton.waitFor({ state: 'visible', timeout: 10000 });
    }
    
    //Select one way Journey
    async selectOneWayJourney() {
        await this.dateOfDepartureInput.click();
        await this.oneWayTab.waitFor({ state: 'visible', timeout: 7000 });
        await this.oneWayRadio.waitFor({ state: 'visible', timeout: 7000 });
        await this.oneWayRadio.click();
        //await this.dateOfDeparture.click();
    }
    
    /**
     * Set departure date
    *@param {string|number} dateOrDays - Date in format 'YYYY-MM-DD' or number of days from now
    *//**
    async setDepartureDate(dateOrDays) {
        const date = typeof dateOrDays === 'number' ? getFutureDate(dateOrDays) : dateOrDays;
        await this.departureDate.fill(date);
        await this.page.waitForLoadState('networkidle'); // 
    }*/
   
    /** Fill origin station - Mardird - Atocha
     * @param {string} station
     */
    async fillOriginStation(station) {
        await this.originStation.fill(station);
        // wait for dropdown options to appear
        await this.page.waitForSelector('li[role="option"]', { timeout: 5000 });
    }

    /** Select droppdown
     * @param {string} optionText
     */
    async selectFromDropdown(optionText) {
        const option = this.page.locator('li[role="option"]', { hasText: optionText }).first();
        await option.click();
        await this.page.waitForLoadState('networkidle'); // Wait for dropdown to close 
    }

    /**
     * Fill destination station - Barcelona - Sants
     * @param {string} station
     */
    async fillDestinationStation(station) {
        await this.destinationStation.fill(station);
        //await this.page.waitForSelector('li[role="option"]', { timeout: 5000 });//wait for dropdown
    }



    /**
     * Click search button
     */
    async clickSearchButton() {
        await this.searchButton.click();
        await this.page.waitForLoadState('networkidle' );
    }

    /**
     * The complete search
     * @param {string} origin
     * @param {string} destination
     * @param {string} date
     */
    async performeSearch(origin, destination, date) {
        await this.selectOneWayJourney();
        await this.fillOriginStation(origin);
        await this.selectFromDropdown('Madrid-Atocha Cercanías');
        await this.fillDestinationStation(destination);
        await this.selectFromDropdown('Barcelona-Sants');
    //    await this.setDepartureDate(date);
        await this.clickSearchButton();
    }
}

File: tests/pages/home_page.py
"""
HomePage - RENFE homepage and search functionality.
"""
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.common.exceptions import TimeoutException
import time
import logging
from .base_page import BasePage

logger = logging.getLogger(__name__)


class HomePage(BasePage):
    """Page object for RENFE homepage."""
    
    # Locators
    DATE_OF_DEPARTURE_INPUT = (By.XPATH, '//input[@id="first-input"]')
    ONE_WAY_RADIO = (By.XPATH, '//label[contains(text(),"Viaje solo ida")]')
    ONE_WAY_TAB = (By.XPATH, '//label[@class="lightpick__label"]')
    ORIGIN_STATION = (By.XPATH, '//input[@id="origin"]')
    DESTINATION_STATION = (By.XPATH, '//input[@id="destination"]')
    DEPARTURE_DATE = (By.XPATH, '//input[@class="rf-daterange-picker-alternative__input"]')
    SEARCH_BUTTON = (By.XPATH, '//button[@type="submit"]')
    ACCEPT_COOKIES_BUTTON = (By.XPATH, '//button[@id="onetrust-accept-btn-handler"]')
    DROPDOWN_OPTION = (By.CSS_SELECTOR, 'li[role="option"]')
    
    def accept_cookies(self):
        """Accept cookie consent banner if present."""
        try:
            if self.is_element_visible(self.ACCEPT_COOKIES_BUTTON, timeout=10):
                logger.info("Cookie banner found, accepting cookies")
                time.sleep(0.5)  # Small delay for stability
                self.click(self.ACCEPT_COOKIES_BUTTON)
                time.sleep(0.5)
                logger.info("Cookies accepted")
        except Exception as e:
            logger.info(f"Cookie banner not present or already accepted: {e}")
    
    def wait_for_page_to_load(self):
        """Wait for the page to be fully loaded."""
        self.wait_for_page_load()
        # Wait for key elements to be visible
        self.wait_for_element(self.DATE_OF_DEPARTURE_INPUT, timeout=10)
        self.wait_for_element(self.ORIGIN_STATION, timeout=10)
        self.wait_for_element(self.DESTINATION_STATION, timeout=10)
        self.wait_for_element(self.SEARCH_BUTTON, timeout=10)
        logger.info("Home page fully loaded")
    
    def select_one_way_journey(self):
        """Select one-way journey type."""
        logger.info("Selecting one-way journey")
        self.click(self.DATE_OF_DEPARTURE_INPUT)
        time.sleep(0.5)
        self.wait_for_element(self.ONE_WAY_TAB, timeout=7)
        self.wait_for_element(self.ONE_WAY_RADIO, timeout=7)
        self.click(self.ONE_WAY_RADIO)
        time.sleep(0.3)
        logger.info("One-way journey selected")
    
    def fill_origin_station(self, station):
        """
        Fill origin station.
        
        Args:
            station: Station name to enter
        """
        logger.info(f"Filling origin station: {station}")
        self.fill(self.ORIGIN_STATION, station)
        time.sleep(0.5)
        # Wait for dropdown options to appear
        self.wait_for_element(self.DROPDOWN_OPTION, timeout=5)
    
    def fill_destination_station(self, station):
        """
        Fill destination station.
        
        Args:
            station: Station name to enter
        """
        logger.info(f"Filling destination station: {station}")
        self.fill(self.DESTINATION_STATION, station)
        time.sleep(0.5)
    
    def select_from_dropdown(self, option_text):
        """
        Select option from dropdown by text.
        
        Args:
            option_text: Text of the option to select
        """
        logger.info(f"Selecting from dropdown: {option_text}")
        # Find the dropdown option containing the text
        dropdown_xpath = f'//li[@role="option" and contains(text(), "{option_text}")]'
        option_locator = (By.XPATH, dropdown_xpath)
        self.wait_for_element(option_locator, timeout=5)
        self.click(option_locator)
        time.sleep(0.5)
        logger.info(f"Selected: {option_text}")
    
    def click_search_button(self):
        """Click the search button."""
        logger.info("Clicking search button")
        self.click(self.SEARCH_BUTTON)
        time.sleep(1)
        self.wait_for_page_load()
        logger.info("Search submitted")
    
    def perform_search(self, origin, destination):
        """
        Perform a complete search.
        
        Args:
            origin: Origin station name
            destination: Destination station name
        """
        self.select_one_way_journey()
        self.fill_origin_station(origin)
        self.select_from_dropdown('Madrid-Atocha Cercanías')
        self.fill_destination_station(destination)
        self.select_from_dropdown('Barcelona-Sants')
        self.click_search_button()
